File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/metadata/mailDriverManager.java
package com.sdkwork.Mail.metadata;

import java.util.List;

public final class MailDriverManager {

  public MailProviderSelection resolveSelection(
      MailProviderSelection.MailProviderSelectionRequest request
  ) {
    return resolveSelection(request, MailProviderCatalog.DEFAULT_mail_PROVIDER_KEY);
  }

  public MailProviderSelection resolveSelection(
      MailProviderSelection.MailProviderSelectionRequest request,
      String defaultProviderKey
  ) {
    return MailProviderSelection.resolveMailProviderSelection(request, defaultProviderKey);
  }

  public MailProviderSupport describeProviderSupport(String providerKey) {
    var official = MailProviderCatalog.getMailProviderByProviderKey(providerKey).isPresent();
    var activation = MailProviderActivationCatalog.getMailProviderActivationByProviderKey(providerKey);

    return MailProviderSupport.createMailProviderSupportState(
        new MailProviderSupport.MailProviderSupportStateRequest(
            providerKey,
            activation
                .map(MailProviderActivationCatalog.MailProviderActivationCatalogEntry::builtin)
                .orElse(false),
            official,
            activation
                .map(MailProviderActivationCatalog.MailProviderActivationCatalogEntry::runtimeBridge)
                .orElse(false)
        )
    );
  }

  public List<MailProviderSupport> listProviderSupport() {
    return MailProviderCatalog.ENTRIES.stream()
        .map(entry -> describeProviderSupport(entry.providerKey()))
        .toList();
  }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-kotlin/src/main/kotlin/com/sdkwork/Mail/metadata/mailDriverManager.kt
package com.sdkwork.Mail.metadata

class MailDriverManager {
    fun resolveSelection(
        request: MailProviderSelectionRequest = MailProviderSelectionRequest(),
        defaultProviderKey: String = MailProviderCatalog.DEFAULT_mail_PROVIDER_KEY,
    ): MailProviderSelection {
        return resolveMailProviderSelection(request, defaultProviderKey)
    }

    fun describeProviderSupport(providerKey: String): MailProviderSupport {
        val official = getMailProviderByProviderKey(providerKey) != null
        val activation = getMailProviderActivationByProviderKey(providerKey)

        return createMailProviderSupportState(
            MailProviderSupportStateRequest(
                providerKey = providerKey,
                builtin = activation?.builtin ?: false,
                official = official,
                registered = activation?.runtimeBridge ?: false,
            )
        )
    }

    fun listProviderSupport(): List<MailProviderSupport> {
        return MailProviderCatalog.entries.map { describeProviderSupport(it.providerKey) }
    }
}
